### DISCUSSION TITLE
InputText: support icon on both sides

### DISCUSSION THREAD

**Discussion by agm1984** [primefaces/community#1500] [PrimeVue]
Recently Primevue updated and lost the ability to have an icon in an input on both left and right.

I have a search input that has a magnifying glass icon on the left side and I was using the right side icon CSS to show an X button that resets the search value. Would be nice to support that with the new IconField component.

**Comment by joaomarcelofm** [comment]


**Reply by agm1984** [reply]
Maybe I wasn't clear. I want the icon on both sides at the same time, like `icon-position="both"`

**Comment by FabianMontoya** [comment]
@agm1984 do you want do something like this?

https://primevue.org/inputgroup/#multiple

![Screenshot 2024-03-25 at 10 58 59](https://github.com/primefaces/community/assets/37662924/f8a21b7c-abc6-4ffa-8ec9-3e95877f5564)
![Screenshot 2024-03-25 at 10 59 17](https://github.com/primefaces/community/assets/37662924/df971a49-4d11-418d-a761-5e614d9507b1)

![Screenshot 2024-03-25 at 10 59 30](https://github.com/primefaces/community/assets/37662924/077654ec-cf25-41ec-b2fc-123802123dd3)

**Reply by agm1984** [reply]
Not quite, I want the search icon to be on the left side of the input and when there is text inside, I want there to be a reset icon on the right side that allows you to clear the search input.

Thats what I had until a recent version of Primevue stopped working.

**Reply by FabianMontoya** [reply]
Do you say that in an older version you can add two icons at the same time for both sides of the `IconField`??

**Reply by FabianMontoya** [reply]
Because, if you dont really need that the both icons appear at the same time, you can do a conditional rendering for the icon that you want to show, and change programmatically the `iconPosition` prop.

But I think it isn't you use case

**Reply by agm1984** [reply]
I didnt use IconField previously, and I might be able to hack it again, but it's not clear currently.

Here is my old code

```
    <span
        :class="['p-input-icon-left', {
            'p-input-icon-right': !!modelValue,
            'w-full': fullWidth,
        }]"
    >
        <i class="icon-search"></i>

        <InputText
            :id="id"
            :model-value="modelValue"
            :placeholder="placeholder"
            :class="{
                'w-48': !fullWidth,
                'w-full': fullWidth,
                'p-invalid': (String(modelValue).length >= 50),
            }"
            maxlength="50"
            @update:model-value="newVal => emit('update:model-value', newVal)"
            @keyup="debouncedSearch"
        />

        <i v-if="!!modelValue" class="">
            <Button
                :id="`${id}-reset`"
                type="button"
                class="p-button-rounded p-button-text p-button-plain !w-8 !h-8 !-mt-2 !-mr-1"
                icon="icon-x"
                @click="handleReset"
            />
        </i>
    </span>
```

**Reply by FabianMontoya** [reply]
Umm, I understand, I think that the solution could be that the `TextInput` has a `cleareable` prop that allow show a cross icon for clear the input, similar as `PasswordInput` works with the eye icon (show/hide the password text)

**Reply by agm1984** [reply]
Our stopgap solution now is to omit the reset button. I also tried a hack solution with absolute position but it didnt work on the first try.

**Reply by agm1984** [reply]
I would be heavily interested in a clearable prop. I think the autocomplete component has it. I might work on this component for a bit today and see if I can hack something.

I'll leave it up to you to decide if any action should take place.

**Reply by FabianMontoya** [reply]
I really don't know how it could look like, but you could try a combination between InputGroup and FieldIcon, to show the option to reset the field.

**Reply by agm1984** [reply]
I'll experiment with InputGroup. Thanks

**Reply by agm1984** [reply]
I just solved it using absolute positioning (tailwind)

```
<script setup>
import debounce from 'lodash.debounce';
import IconField from 'primevue/iconfield';
import InputIcon from 'primevue/inputicon';
import InputText from 'primevue/inputtext';
import Button from 'primevue/button';

defineProps({
    id: {
        type: String,
        required: false,
        default: 'search',
    },

    modelValue: {
        type: [String, Number],
        required: true,
    },

    placeholder: {
        type: String,
        required: false,
        default: 'Search',
    },

    fullWidth: {
        type: Boolean,
        requried: false,
        default: false,
    },
});

const emit = defineEmits([
    'update:model-value',
    'search',
]);

const handleSearch = () => {
    emit('search');
};

const debouncedSearch = debounce(handleSearch, 500);

const handleReset = () => {
    emit('update:model-value', '');
    emit('search');
};
</script>

<template>
    <div :class="['relative', { 'w-full': fullWidth }]">
        <IconField
            :class="['', { 'w-full': fullWidth }]"
            icon-position="left"
            :pt="{ class: fullWidth && 'w-full' }"
        >
            <InputIcon class="icon-search"></InputIcon>

            <InputText
                :id="id"
                :model-value="modelValue"
                :placeholder="placeholder"
                :class="{
                    'w-48': !fullWidth,
                    'w-full': fullWidth,
                }"
                @update:model-value="newVal => emit('update:model-value', newVal)"
                @keyup="debouncedSearch"
            />
        </IconField>

        <Button
            v-if="!!modelValue"
            :id="`${id}-reset`"
            type="button"
            class="absolute top-[11px] right-[8px] p-button-rounded p-button-text p-button-plain !w-8 !h-8 !-mt-2 !-mr-1"
            icon="icon-x"
            @click="handleReset"
        />
    </div>
</template>

```

**Comment by matthew-dean** [comment]
I came here with the same issue and would love a solution. I'd either like:

1. The `InputText` to have a `showClear` option (like DropDown), such that I can have a Search box with both a search icon (on left) and a clear icon (on right). This is a common pattern and an InputGroup is not the typical UI for this scenario.
2. I'd love it if `iconPosition` went away entirely and icon positions were determined by source order.

i.e.:
```vue
<IconField>
    <InputIcon><IconSearch /></InputIcon>
    <InputText v-model="search" />
    <InputIcon><IconClear /></InputIcon>
</IconField>
```
To me, the above is the friendliest, most intuitive API. It's counter-intuitive and seems wholly unnecessary that the icon MUST appear first even if appearing on the right.

In fact, I see no legitimate reason why the number of icons needs in a field to have any limit. Like InputGroup, they could stack on either side:
```vue
<IconField>
    <InputIcon><IconSearch /></InputIcon>
    <InputText v-model="search" />
    <InputIcon><IconClear /></InputIcon>
    <InputIcon><IconValidationFlag /></InputIcon>
</IconField>
```

One of my biggest complaints about PrimeVue (even though I appreciate the library in general and that it's available!) is that the API is wildly, wildly inconsistent, and the difference in pattern between `IconGroup` and `IconField` is a prime example. Even their naming is different, with one describing the group and one... describing the icon? Despite the fact it has an input in it?

_(As an aside, it often feels like each PrimeVue component was made by a different engineer in isolation who had no insight into any component made before it. Not only is the naming for props and components different, but even the methods for controlling, say, visibility of an element are spectacularly different. It's one reason why I wanted to abstract each PrimeVue component with another Vue wrapper, so I could expose a consistent API to our FE engineers for better DX, but this turned out to be near-impossible because of PrimeVue's unique component structure.)_

**Reply by agm1984** [reply]
I posted a solution above in the reply thread. Check it out.
I haven't tested it against text appearing and running behind the icon, but you could put a white background behind the X button if you need.
